### PR TITLE
Update build scripts

### PR DESCRIPTION
**Pull request by RussKie**
* Disable these:
   image
   image

* Simply build structure removing reset-commit steps that were responsible for setting original PR commit's hash. Override `$(GitCommit)` and `$(GitSha)` MSBuild properties instead.
For example: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45871126
  image
  image
  image

### CODE REVIEW

File: GitExtensions/Project.Loc.targets [hunk 1]
```diff
@@ -68,7 +68,7 @@
           ContinueOnError="false" />
 
     <!-- 5. Verify no changes to English*.xlf -->
-    <Exec Command="git add *.xlf --ignore-errors"
+    <Exec Command="cmd /c git add *.xlf --ignore-errors > /dev/null 2>&1"
```
Comment (gerhardol) [inline GitExtensions/Project.Loc.targets:71]: The change hides other errors too, suggest to suppress autocrlf instead.
(Have not tried this in a build.)
```suggestion
    <Exec Command="git -c core.autocrlf=false add *.xlf --ignore-errors"
```

**Review by gerhardol — COMMENTED** [review]
+0
Do not oppose a merge, but prefer not to hide all errors

File: scripts/set_version_to.py [hunk 2]
```diff
@@ -54,7 +54,7 @@
                     data[1] = args.text
                     line = '"'.join(data)
             o += line
-        outfile = open(filename, "w")
+        outfile = open(filename, "w", newline='\n')
```
Comment (mstv) [inline scripts/set_version_to.py:57]: 👍 not attempting to cure the symptom any longer

File: appveyor.experimental.yml [hunk 3]
```diff
@@ -48,7 +48,7 @@ install:
 - cmd: echo /logger:"%ProgramFiles%\AppVeyor\BuildAgent\Appveyor.MSBuildLogger.dll">> Directory.Build.rsp
 - cmd: |-
     cd scripts
-    C:\\Python35\\python set_version_to.py -v %APPVEYOR_BUILD_VERSION% -t %APPVEYOR_BUILD_VERSION%
+    C:\\Python311\\python set_version_to.py -v %APPVEYOR_BUILD_VERSION% -t %APPVEYOR_BUILD_VERSION%
```
Comment (mstv) [inline appveyor.experimental.yml:51]: This file is pretty outdated. It should be sync'ed with appveyor.yml. I committed a quick first draft: 0fa03a24bc8e6e69307c4314c9da017954601d43 in parallel.

**Review by gerhardol — APPROVED** [review]
The update commit by mstv (as in my suggestion) resolves the printout.

The description of the PR did not address the description.

```
EXEC : warning : in the working copy of 'CommonAssemblyInfo.cs', CRLF will be replaced by LF the next time Git touches it [C:\projects\gitextensions\BugReporter\BugReporter.csproj]
```
I fixed the warning (by writing as lf).
Also updated Python (Py35 may well be removed for some security issue at one point), used the latest version available in AppVeyor.

The warning for other changed files by users (the first example) is not handled.
As text=auto in .gitattributes, crlf should be converted to lf at commit, so these files are incorrectly committed

File: scripts/set_version_to.py (hunk 2 above)
Comment (mstv) [reply scripts/set_version_to.py:57]: Tough it is not the source of
```
warning: in the working copy of 'CommonAssemblyInfoExternals.cs', LF will be replaced by CRLF the next time Git touches it
warning: in the working copy of 'Externals/NetSpell.SpellChecker/Properties/AssemblyInfo.cs', LF will be replaced by CRLF the next time Git touches it
```

**Review by mstv — APPROVED** [review]
Outdated translations are still detected 👍
https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45864645

File: GitExtensions/Project.Loc.targets [hunk 4]
```diff
@@ -68,7 +68,7 @@
           ContinueOnError="false" />
 
     <!-- 5. Verify no changes to English*.xlf -->
-    <Exec Command="git add *.xlf --ignore-errors"
+    <Exec Command="git -c core.autocrlf=false add *.xlf --ignore-errors"
```
Comment (RussKie) [inline GitExtensions/Project.Loc.targets:71]: Should it be "autocrlf=input"?

Comment (RussKie) [reply GitExtensions/Project.Loc.targets:71]: Same for the appveyor.yml

File: CommonAssemblyInfo.cs [hunk 5]
```diff
@@ -10,7 +10,7 @@
 [assembly: AssemblyConfiguration("")]
 [assembly: AssemblyCompany("Git Extensions")]
 [assembly: AssemblyProduct("Git Extensions")]
-[assembly: AssemblyCopyright("Copyright © 2008-2019 Git Extensions Team")]
+[assembly: AssemblyCopyright("Copyright © 2008-2023 Git Extensions Team")]
```
Comment (RussKie) [inline CommonAssemblyInfo.cs:13]: This is unrelated

File: appveyor.experimental.yml (hunk 3 above)
Comment (RussKie) [reply appveyor.experimental.yml:51]: Great!

File: CommonAssemblyInfo.cs (hunk 5 above)
Comment (gerhardol) [reply CommonAssemblyInfo.cs:13]: Remains from changing the format, can revert

File: GitExtensions/Project.Loc.targets (hunk 4 above)
Comment (gerhardol) [reply GitExtensions/Project.Loc.targets:71]: With `input` the file is not changed when checking out but when adding to index.
In this case, the setting is used when adding to the index, so effectively same as true. 
See comment to mstv above.

File: scripts/set_version_to.py (hunk 2 above)
Comment (gerhardol) [reply scripts/set_version_to.py:57]: As text=auto in .gitattributes, .cs are converted crlf<->lf
Maybe setting .xlf as text instead?

**Comment by RussKie** [comment]
There are still warnings which I'd like to remove:
image

**Comment by RussKie** [comment]
The latest @gerhardol's tweaks still result in warnings:
image

**Comment by gerhardol** [comment]
Adjusted some more, adding files to .gitattributes to set line endings andtext properties.
text files are checked out as crlf and committed as lf (as text=auto). I do not want to change that setup and add the files to the index after updating version but before building, to avoid warnings.

**Review by mstv — DISMISSED** [review]
👍 nothing red anymore

**Comment by RussKie** [comment]
Awesome! Thank you folks

File: appveyor.yml [hunk 6]
```diff
@@ -78,7 +82,7 @@ build_script:
     if ($isBuildingPR) {
         git config user.email "gitextensions@github.com"
         git config user.name "Git Extensions"
-        git -c core.autocrlf=false commit -a -m "restore merge commit" | Out-Null
+        git -c core.autocrlf=false commit -a -m "commit version files"
         # ignore warning about line-endings conversion
         $global:LASTEXITCODE = 0
```
Comment (RussKie) [inline appveyor.yml:87]: This just occurred to me - we're doing all of these gimmicks because we want to set the PR commit's hash, right? Why not set override the hash explicitly instead of resetting and recommitting?

**Comment by RussKie** [comment]
I simplified the build script overriding commit hash supplied by GitInfo package with the hash supplied by AppVeyor. 

For example: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45871126

image
image
image

File: appveyor.yml [hunk 7]
```diff
@@ -48,43 +48,36 @@ install:
...
+    if ($LastExitCode -ne 0) { $host.SetShouldExit($LastExitCode) }
 - ps: |
     # build VC++
     dotnet build .\scripts\native.proj -c Release --verbosity q --nologo /bl:.\artifacts\log\native.binlog
     if ($LastExitCode -ne 0) { $host.SetShouldExit($LastExitCode) }
 - ps: |
     # build .NET
-    dotnet build -c Release --verbosity q --nologo /bl:.\artifacts\log\build.binlog /p:ContinuousIntegrationBuild=true
+    $cmd = "dotnet build -c Release --verbosity q --nologo /bl:.\artifacts\log\build.binlog $buildArgs"
+    iex $cmd
```
Comment (mstv) [inline appveyor.yml:77]:
```suggestion
    Invoke-Expression $cmd
```
Why is it needed at all?

File: appveyor.yml [hunk 8]
```diff
@@ -48,43 +48,36 @@ install:
...
     cd ..
     git --version
 
 
 # to run your custom scripts instead of automatic MSBuild
 build_script:
 - ps: |
-    $isBuildingPR = $false;
-    # if building a temporary merge with master, soft reset to the PR commit so the build contains the PR's hash instead of the merge-commit's hash
+    $buildArgs = '/p:ContinuousIntegrationBuild=true';
```
Comment (mstv) [inline appveyor.yml:59]: nit:
```suggestion
    $buildArgs = '/p:ContinuousIntegrationBuild=true'
```

File: appveyor.yml [hunk 9]
```diff
@@ -48,43 +48,36 @@ install:
...
+    # if building a temporary merge with master, override the hash with the PR commit's so the build contains the PR's hash instead of the merge-commit's hash
     if ($env:APPVEYOR_PULL_REQUEST_HEAD_COMMIT -and ($env:APPVEYOR_REPO_COMMIT -ne $env:APPVEYOR_PULL_REQUEST_HEAD_COMMIT))
     {
-        git reset --soft "$env:APPVEYOR_PULL_REQUEST_HEAD_COMMIT" --
-        $isBuildingPR = $true;
+        $gitCommit = $(git rev-parse --short $env:APPVEYOR_PULL_REQUEST_HEAD_COMMIT)
+        $buildArgs +=  " /p:GitCommit=$gitCommit /p:GitSha=$($env:APPVEYOR_PULL_REQUEST_HEAD_COMMIT)"
```
Comment (mstv) [inline appveyor.yml:64]: nit:
```suggestion
        $buildArgs += " /p:GitCommit=$gitCommit /p:GitSha=$($env:APPVEYOR_PULL_REQUEST_HEAD_COMMIT)"
```

File: appveyor.yml [hunk 10]
```diff
@@ -100,7 +93,8 @@ test_script:
 after_test:
 - ps: |
     Write-Host "Preparing build artifacts..."
-    dotnet publish -c Release --no-build /bl:.\artifacts\log\publish.binlog /p:ContinuousIntegrationBuild=true
+    $cmd = "dotnet publish -c Release --no-build /bl:.\artifacts\log\publish.binlog $buildArgs"
+    iex $cmd
```
Comment (mstv) [inline appveyor.yml:97]: ditto

File: appveyor.yml [hunk 11]
```diff
@@ -48,43 +48,36 @@ install:
...
-        git config user.name "Git Extensions"
-        git -c core.autocrlf=false commit -a -m "restore merge commit" | Out-Null
-        # ignore warning about line-endings conversion
-        $global:LASTEXITCODE = 0
-    }
 - ps: |
     # Verify that new strings (if any) have been processed and ready for localisation
-    # it would be nice to run '.\cibuild.cmd -loc -logFileName localise.binlog /p:NoBuild=true' but it doesn't work without `-build` switch :\
     Push-Location .\GitExtensions
     dotnet msbuild /p:Configuration=Release /t:_UpdateEnglishTranslations /p:RunTranslationApp=true /p:ContinuousIntegrationBuild=true /v:m /bl:..\artifacts\log\localise.binlog
```
Comment (mstv) [inline appveyor.yml:82]:
```suggestion
    dotnet msbuild /p:Configuration=Release /t:_UpdateEnglishTranslations /p:RunTranslationApp=true $buildArgs /v:m /bl:..\artifacts\log\localise.binlog
```

File: GitExtensions/Project.Publish.targets [hunk 12]
```diff
@@ -229,9 +229,9 @@
 
       <!-- Resolve the output file -->
       <_PublishPortableVersionSuffix>-$(CurrentBuildVersion.ToString())</_PublishPortableVersionSuffix>
+
+      <!-- This property comes from GitInfo package, but can be overriden on AppVeyor -->
       <_PublishPortableCommitHashSuffix Condition="'$(GitCommit)' != ''">-$(GitCommit)</_PublishPortableCommitHashSuffix>
-      <_PublishPortableCommitHashSuffix Condition="'$(env:APPVEYOR_REPO_COMMIT)' != ''">-$(env:APPVEYOR_REPO_COMMIT)</_PublishPortableCommitHashSuffix>
-      <_PublishPortableCommitHashSuffix Condition="'$(env:APPVEYOR_PULL_REQUEST_HEAD_COMMIT)' != ''">-$(env:APPVEYOR_PULL_REQUEST_HEAD_COMMIT)</_PublishPortableCommitHashSuffix>
```
Comment (mstv) [inline GitExtensions/Project.Publish.targets:234]: Env `GitCommit` is not set for builds of `master`.
Will the artifact names still contain the hash?

**Review by gerhardol — APPROVED** [review]

File: GitExtensions/Project.Publish.targets (hunk 12 above)
Comment (RussKie) [reply GitExtensions/Project.Publish.targets:234]: `$(GitCommit)` is set by [GitInfo](https://github.com/devlooped/GitInfo).

File: appveyor.yml (hunk 11 above)
Comment (RussKie) [reply appveyor.yml:82]: This wouldn't work as you expect it - `$buildArgs` don't get interpreted as input parameters:
image
https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45871046

So, the whole command needs to be encoded as a string and then executed.

Comment (RussKie) [reply appveyor.yml:82]: I specifically left this command as is - the translation verification step does not perform a build and it does not require commit hashes.

File: GitExtensions/Project.Publish.targets (hunk 12 above)
Comment (RussKie) [reply GitExtensions/Project.Publish.targets:234]: Here's a non-PR build: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45874141

Comment (mstv) [reply GitExtensions/Project.Publish.targets:234]: The non-PR builds do not upload the `.msi`.

Comment (RussKie) [reply GitExtensions/Project.Publish.targets:234]: That's correct. Only a PR can conditionally upload msi
https://github.com/gitextensions/gitextensions/blob/24b715979fee11688a6cfba6f12d7d5c68c77036/appveyor.yml#L99-L103

Comment (mstv) [reply GitExtensions/Project.Publish.targets:234]: But it has been reversed somehow:
branch build: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45874141/artifacts
merge build PR: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45874136/artifacts
**master build**: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45874709/artifacts

**Comment by gerhardol** [comment]
The solution can be slightly simplified as there is no temporary commit, removing the adding to index.
e6212fa4c8f4061c56a495e41fc3c3b9c6f9ea2f

File: GitExtensions/Project.Publish.targets (hunk 12 above)
Comment (RussKie) [reply GitExtensions/Project.Publish.targets:234]: > branch build: [ci.appveyor.com/project/gitextensions/gitextensions/builds/45874141/artifacts](https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45874141/artifacts)

Doesn't have msi - expected

> merge build PR: [ci.appveyor.com/project/gitextensions/gitextensions/builds/45874136/artifacts](https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45874136/artifacts)

The msi publishing was enabled for the PR, so the msi is present - expected

> **master build**: [ci.appveyor.com/project/gitextensions/gitextensions/builds/45874709/artifacts](https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45874709/artifacts)

Doesn't have msi - expected

What am I missing?

**Comment by RussKie** [comment]
> The solution can be slightly simplified as there is no temporary commit, removing the adding to index.
> [e6212fa](https://github.com/gitextensions/gitextensions/commit/e6212fa4c8f4061c56a495e41fc3c3b9c6f9ea2f)

If it works - 👍 
I tried removing the step, but was still observing warnings... ¯\\_(ツ)\_/¯

**Comment by gerhardol** [comment]
> > The solution can be slightly simplified as there is no temporary commit, removing the adding to index.
> > [e6212fa](https://github.com/gitextensions/gitextensions/commit/e6212fa4c8f4061c56a495e41fc3c3b9c6f9ea2f)
> 
> If it works - 👍 I tried removing the step, but was still observing warnings... ¯\_(ツ)_/¯

Also writing the updated version files as LF, to not get warnings when building. (Writing LF gave warnings when committing previously.)

When there are other changes...

File: GitExtensions/Project.Publish.targets (hunk 12 above)
Comment (RussKie) [reply GitExtensions/Project.Publish.targets:234]: We don't publish msi from dev builds, only for release builds. It's been like that for a very long time.

Comment (mstv) [reply GitExtensions/Project.Publish.targets:234]: Yes, I have mixed it up with builds of a release branch: https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45820457/artifacts

**Comment by RussKie** [comment]
@gerhardol it looks like something got broken - the master is failing to build tripping at `git -c core.autocrlf=false add -A` command. And it started [here](https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45888832). And curiously the [previous build](https://ci.appveyor.com/project/gitextensions/gitextensions/builds/45874709) was green....
image

The release/4.0 branch is green with the same command.

Any thoughts?
